src/clr-angular/data/datagrid/providers/columns.service.ts

```diff
@@ -19,6 +19,37 @@ export class ColumnsService {
     return this.columnStates.filter(state => state.hideable).length > 0;
   }
 
+  get orderOfLastVisible(): number {
+    return Math.max(...this.columnStates.filter(state => !state.hidden).map(state => state.order));
+  }
+
+  get orderOfFirstVisible(): number {
+    return Math.min(...this.columnStates.filter(state => !state.hidden).map(state => state.order));
+  }
+
+  // First and last visible columns may change in 3 different cases:
+  // -  When columns are reordered
+  // -  When the current first/last gets hidden state
+  // -  When the column at first/last gets removed or a new column added at first/last index
+  private _isFirstVisibleChanged: Subject<void> = new Subject<void>();
+  private _isLastVisibleChanged: Subject<void> = new Subject<void>();
```

Comment:
As I commented elsewhere, I'm not sure we need two different Subjects for this. I'm also not sure if this is the right place, it seems different from what the service core purpose is right now, should this go into the reorder service?